models/sipnet/R/write.events.SIPNET.R

```diff
@@ -0,0 +1,105 @@
+## TODO:
+## - integrate call into write.configs.SIPNET
+## - parameterize planting allocation fractions
+## - make sure files are written in correct output directory
+## - map crops associated w/ planting and harvest --> PFTs; this will need to be handled separate from events.in
+#' Write SIPNET events.in files from a PEcAn events.json
+#'
+#' Reads a single PEcAn events.json containing one or more site objects and
+#' writes one SIPNET `events.in` file per site. Events are translated according to [SIPNET's `events.in`
+#' specification](https://pecanproject.github.io/sipnet/parameters/#agronomic-events).
+#' The writer expects inputs to already match the PEcAn MVP schema v0.1.0 naming and units where applicable.
+#'
+#' @details
+#' - Supported `event_type` values: `tillage`, `planting`, `fertilization`,
+#'   `irrigation`, `harvest`.
+#' - Units translated from PEcAn standard_vars to SIPNET events.in specification:
+#'   `kg/m^2` to `g/m^2`; irrigation `amount_mm` to `cm`.
+#' - Planting allocation uses fixed internal parameters. Future work should use the same values
+#'   that are written to `sipnet.parms` (e.g. after integrating this into `write.configs.SIPNET`)
+#'
+#' @param events_json character. Path to an `events.json` file containing an
+#'   array of site objects with `site_id`, optional `pft`, and `events`.
+#' @param outdir character. Output directory where per-site `events-<site>.in`
+#'   files are written.
+#'
+#' @return Invisibly, a vector of files written.
+#'
+#' @examples
+#' # Example with two events for a single site
+#' tmp <- withr::local_tempfile(fileext = ".json")
+#' site <- list(
+#'     site_id = "EX1",
+#'     events = list(
+#'         list(event_type = "tillage", date = "2022-02-04", tillage_eff_0to1 = 0.2),
+#'         list(event_type = "planting", date = "2022-02-19", leaf_c_kg_m2 = 0.01)
+#'     )
+#' )
+#' jsonlite::write_json(list(site), tmp, auto_unbox = TRUE)
+#' outdir <- withr::local_tempdir()
+#' files <- write.events.SIPNET(tmp, outdir)
+#' files
+#'
+#' @export
+write.events.SIPNET <- function(events_json, outdir) {
+    x <- jsonlite::fromJSON(events_json, simplifyVector = FALSE)
+    site_objs <- if (!is.null(x$site_id)) list(x) else x
+    files_written <- vector()
+
+    leafAllocation <- 0.50
+    woodAllocation <- 0.15
+    fineRootAllocation <- 0.10
+    coarseRootAllocation <- 0.25
+
+    # Unit conversion helpers
+    kg2g <- as.numeric(PEcAn.utils::ud_convert(1, "kg", "g")) # 1000
+    mm2cm <- as.numeric(PEcAn.utils::ud_convert(1, "mm", "cm")) # 0.1
+
+    # For each site, build event time series and write file
+    for (site in site_objs) {
+        sid <- site$site_id
+        evs <- site$events
+        # Order by date and build lines
+        dates <- as.Date(vapply(evs, function(e) e$date, character(1)))
+        ord <- order(dates)
```

Comment:
If I'm following the generate_events code in #3621 right, we impose ordering at json build time. Is it necessary to enforce it at both ends / does checking here mean we could save steps upstream by allowing events to be out-of-order until here?